extension/src/state_aggregate.rs

```diff
@@ -71,6 +110,85 @@ pub mod toolkit_experimental {
             let end = record.state_end as usize;
             &self.states_as_str()[beg..end]
         }
+
+        #[allow(clippy::unnecessary_unwrap)] // following clippy would make this much more convoluted 
+        pub(super) fn interpolate(
+            &self,
+            interval_start: i64,
+            interval_len: i64,
+            prev: Option<StateAgg>,
+            next: Option<StateAgg>,
+        ) -> StateAgg {
+            if self.durations.is_empty() {
+                pgx::error!("unable to interpolate interval on state aggregate with no data");
+            }
+
+            let mut states = std::str::from_utf8(self.states.as_slice())
+                .unwrap()
+                .to_string();
+            let mut durations: Vec<DurationInState> =
+                self.durations.iter().collect();
+
+            let first = if interval_start < self.first_time && prev.is_some() {
+                let prev = prev.unwrap();
+                if prev.last_state < prev.durations.len() as u32 {
+                    let start_interval = self.first_time - interval_start;
+                    let start_state =
+                        prev.state_str(&prev.durations.as_slice()[prev.last_state as usize]);
+                    match durations.iter_mut().find(|x| {
+                        states[x.state_beg as usize..x.state_end as usize].eq(start_state)
+                    }) {
+                        Some(dis) => dis.duration += start_interval,
+                        None => {
+                            durations.push(DurationInState {
+                                duration: start_interval,
+                                state_beg: states.len() as u32,
+                                state_end: (states.len() + start_state.len() - 1) as u32,
+                            });
+                            states += start_state;
+                        }
+                    };
+
+                    Record {
+                        state: start_state.to_string(),
+                        time: interval_start,
+                    }
+                } else {
+                    pgx::error!("unable to interpolate interval on state aggregate where previous agg has no data")
+                }
+            } else {
+                Record {
+                    state: self
+                        .state_str(&self.durations.as_slice()[self.first_state as usize])
+                        .to_string(),
+                    time: self.first_time,
+                }
+            };
+
+            let last = if interval_start + interval_len > self.last_time && next.is_some() {
```

Comment:
Hmm, because state_agg uses LOCF interpolation, all we really care about is whether or not the next value exists.  This could be a bool, though I do know that where we use it it's easier to just pass through the argument postgres is giving us.  However, that might be assuming too much about how this is used.